src/net/sourceforge/kolmafia/textui/langserver/FilesMonitor.java

```diff
@@ -18,6 +20,44 @@ protected FilesMonitor(final AshLanguageServer parent) {
     this.parent = parent;
   }
 
+  /** Updates the information we have about a file based on information from the client. */
+  public void updateFile(final File file, final String text, final int version) {
+    if (file == null) {
+      return;
+    }
+
+    synchronized (this.parent.scripts) {
+      final Script script = this.getScript(file);
+
+      final List<Script.Handler> handlers = this.findHandlers(file);
+
+      if (handlers.size() > 0
+          && ((text != null && script.version >= version)
+              || (script.text == null && text == null))) {
+        // We already have a working handler using an
+        // up-to-date version of that file.
+        return;
+      }
+
+      if (script.text == null || script.version < version || script.text != null && text == null) {
+        script.text = text;
+        script.version = version;
+      }
+
+      if (handlers.size() == 0) {
+        // make a new handler
+        handlers.add(script.makeHandler());
```

Comment:
If this line is straightforward to write a test for, would you mind?

Additionally, does creating a new handler do the equivalent of "refreshParsing"?

Reply:
Not quite.
First, both submit a different boolean to `Handler.parseFile(initialParsing)`.
Second, `makeHandler` also additionally causes the Handler to get registered for future use.